python/ray/util/sgd/tests/test_torch_failure.py

```diff
@@ -0,0 +1,156 @@
+        return
+
+    def single_loader(config):
+        dataset = LinearDataset(2, 5, size=1000000)
+        return DataLoader(dataset, batch_size=config.get("batch_size", 32))
+
+    step_with_fail = gen_step_with_fail(1)
+
+    TestOperator = TrainingOperator.from_creators(
+        model_creator,
+        optimizer_creator,
+        single_loader,
+        loss_creator=lambda config: nn.MSELoss())
+    with patch.object(RemoteWorkerGroup, "_train", step_with_fail):
+        trainer1 = TorchTrainer(
+            training_operator_cls=TestOperator,
+            config={"batch_size": 100000},
+            use_local=use_local,
+            num_workers=2)
+
+        original_state_dict = trainer1.state_dict()
+
+        @ray.remote
+        def try_test():
+            import time
+            time.sleep(100)
+
+        try_test.remote()
+        trainer1.train(max_retries=1)
+        assert len(trainer1.worker_group.remote_workers) == 1
+        assert trainer1.state_dict() == original_state_dict
+
+        trainer1.shutdown()
+
+
+@pytest.mark.parametrize("use_local", [False, True])
+def test_fail_twice(ray_start_2_cpus, use_local):  # noqa: F811
+    if not dist.is_available():
+        return
+
+    def single_loader(config):
+        dataset = LinearDataset(2, 5, size=1000000)
+        return DataLoader(dataset, batch_size=config.get("batch_size", 32))
+
+    step_with_fail = gen_step_with_fail(2)
+
+    TestOperator = TrainingOperator.from_creators(
+        model_creator,
+        optimizer_creator,
+        single_loader,
+        loss_creator=lambda config: nn.MSELoss())
+
+    with patch.object(RemoteWorkerGroup, "_train", step_with_fail):
+        trainer1 = TorchTrainer(
+            training_operator_cls=TestOperator,
+            config={"batch_size": 100000},
+            use_local=use_local,
+            num_workers=2)
+
+        # MAX RETRIES SHOULD BE ON BY DEFAULT
+        trainer1.train()
+        trainer1.shutdown()
+
+
+@pytest.mark.parametrize("use_local", [False, True])
+def test_fail_with_recover(ray_start_2_cpus, use_local):  # noqa: F811
+    print(locals())
+    if not dist.is_available():
+        return
+
+    def single_loader(config):
+        dataset = LinearDataset(2, 5, size=1000000)
+        return DataLoader(dataset, batch_size=config.get("batch_size", 32))
+
+    step_with_fail = gen_step_with_fail(3)
+    TestOperator = TrainingOperator.from_creators(
+        model_creator,
+        optimizer_creator,
+        single_loader,
+        loss_creator=lambda config: nn.MSELoss())
+    with patch.object(RemoteWorkerGroup, "_train", step_with_fail):
+        trainer1 = TorchTrainer(
+            training_operator_cls=TestOperator,
+            config={"batch_size": 100000},
+            timeout_s=5,
+            use_local=use_local,
+            num_workers=2)
+
+        with pytest.raises(RuntimeError):
+            trainer1.train(max_retries=1)
+
+        trainer1.shutdown(force=True)
```

Comment:
you need to add a ifmain block underneath (this isn't running on travis right now)

Reply:
done